File: task3.py
import matplotlib.pyplot as plt

# 1. Read the page references from a file
with open('refs.txt', 'r') as f:
    pages = [int(x) for x in f.read().split()]

total_refs = len(pages)
frames_range = range(1, 11) # Testing from 1 to 10 page frames
faults_per_1000 = []

# 2. Simulate the aging algorithm
for num_frames in frames_range:
    memory = {} # Keeps track of pages and their 'age' counters
    faults = 0
    
    for page in pages:
        # Step A: Age all pages currently in memory by shifting bits right
        for p in memory:
            memory[p] >>= 1
            
        # Step B: Check for a page fault
        if page not in memory:
            faults += 1
            # If memory is full, find the smallest counter and kick that page out
            if len(memory) >= num_frames:
                evict = min(memory, key=memory.get)
                del memory[evict]
            # Bring the new page in
            memory[page] = 0
            
        # Step C: Mark the current page as recently used (set top bit to 1)
        memory[page] |= 128
        
    # Calculate the required metric
    faults_per_1000.append((faults / total_refs) * 1000)

# 3. Plot the results
plt.plot(frames_range, faults_per_1000, marker='o')
plt.title('Aging Algorithm Performance')
plt.xlabel('Number of Page Frames')
plt.ylabel('Page Faults per 1000 References')
plt.show()

File: task3.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <unordered_map>
#include <iomanip>

using namespace std;

int main() {
    // 1. Read the page references
    ifstream file("refs.txt");
    if (!file.is_open()) {
        cerr << "Error: Could not open refs.txt." << endl;
        return 1;
    }

    vector<int> pages;
    int page_num;
    while (file >> page_num) {
        pages.push_back(page_num);
    }
    file.close();

    int total_refs = pages.size();
    if (total_refs == 0) {
        cerr << "Error: The references file is empty." << endl;
        return 1;
    }

    // NEW: Open a CSV file to save our graph data
    ofstream csv_file("results.csv");

    cout << "Frames | Faults per 1000 References" << endl;
    cout << "-----------------------------------" << endl;

    // 2. Simulate the aging algorithm
    for (int num_frames = 1; num_frames <= 10; ++num_frames) {
        unordered_map<int, uint8_t> memory;
        int faults = 0;

        for (int page : pages) {
            // Step A: Age pages
            for (auto& pair : memory) {
                pair.second >>= 1;
            }

            // Step B: Page fault check
            if (memory.find(page) == memory.end()) {
                faults++;
                // Evict if full
                if (memory.size() >= num_frames) {
                    auto evict_it = memory.begin();
                    for (auto it = memory.begin(); it != memory.end(); ++it) {
                        if (it->second < evict_it->second) {
                            evict_it = it;
                        }
                    }
                    memory.erase(evict_it);
                }
                memory[page] = 0;
            }
            // Step C: Mark as used
            memory[page] |= 128;
        }

        // Calculate metric
        double faults_per_1000 = (static_cast<double>(faults) / total_refs) * 1000.0;

        // Print to terminal
        cout << setw(6) << num_frames << " | " << fixed << setprecision(2) << faults_per_1000 << endl;

        // NEW: Save the exact same numbers to the CSV file
        csv_file << num_frames << "," << faults_per_1000 << "\n";
    }

    csv_file.close(); // Close the CSV file
    return 0;
}